---
title: "20250226_PieCharts"
output: html_document
date: "2025-02-26"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load libraries

```{r}
library(ChIPseeker)
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
txdb <- TxDb.Mmusculus.UCSC.mm10.ensGene
library(clusterProfiler)

library(ChIPseeker)
library(GenomicRanges)
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene)
library(org.Dm.eg.db)
library(rtracklayer)
library(ggplot2)
library(dplyr)
setwd("/data/jenuwein/processing/05.Alecia/code/AleciasProject/")
```

##Plot a generic pie chart

```{r, echo=FALSE}
peaks <- readPeakFile("/data/jenuwein/processing/05.Alecia/code/AleciasProject/INTS11_mergedPeaks.bed")

peakAnno <- annotatePeak(peaks, tssRegion=c(-3000, 3000),
                         TxDb=txdb, annoDb="TxDb.Mmusculus.UCSC.mm10.ensGene")

plotAnnoPie(peakAnno)


# Convert to data frames
peakAnno_df <- as.data.frame(peakAnno)

# Initialize customCategory column
peakAnno_df$customCategory <- peakAnno_df$annotation

##################
#extract only the distal intergenic peaks that are not annotated by anything else
#peakAnno_distal<-peakAnno_df[peakAnno_df$annotation == "Distal Intergenic",]


###################
#Everything other the distal intergenenic will be annotated as Genes
peakAnno_distal <- peakAnno_df %>%
  mutate(customCategory = ifelse(customCategory != "Distal Intergenic", "Genes", customCategory))

LINE_repeats <- import.bed("/data/jenuwein/processing/05.Alecia/code/LINEs.5UTR.filtered.bed")

peaks_gr <- GRanges(seqnames = peakAnno_distal$seqnames, ranges = IRanges(peakAnno_distal$start, peakAnno_distal$end))

# Overlap with repeats only for peaks not annotated as LINEs
repeats_gr <- GRanges(seqnames = seqnames(LINE_repeats), ranges = IRanges(start(LINE_repeats), end(LINE_repeats)))
repeat_hits <- findOverlaps(peaks_gr, repeats_gr)
peakAnno_distal$customCategory[queryHits(repeat_hits)] <- "LINEs"

#SINEs
SINE_repeats <- import.bed("/data/jenuwein/processing/05.Alecia/code/SINE.mm10.bed")
repeats_gr <- GRanges(seqnames = seqnames(SINE_repeats), ranges = IRanges(start(SINE_repeats), end(SINE_repeats)))
repeat_hits <- findOverlaps(peaks_gr, repeats_gr)
peakAnno_distal$customCategory[queryHits(repeat_hits)] <- "SINEs"

#LTR
LTR_repeats <- import.bed("/data/jenuwein/processing/05.Alecia/code/LTR.mm10.bed")
repeats_gr <- GRanges(seqnames = seqnames(LTR_repeats), ranges = IRanges(start(LTR_repeats), end(LTR_repeats)))
repeat_hits <- findOverlaps(peaks_gr, repeats_gr)
peakAnno_distal$customCategory[queryHits(repeat_hits)] <- "LTRs"

#GSAT
GSAT_repeats <- import.bed("/data/jenuwein/processing/05.Alecia/code/GSAT.mm10.bed")
repeats_gr <- GRanges(seqnames = seqnames(GSAT_repeats), ranges = IRanges(start(GSAT_repeats), end(GSAT_repeats)))
repeat_hits <- findOverlaps(peaks_gr, repeats_gr)
peakAnno_distal$customCategory[queryHits(repeat_hits)] <- "MSR"



peakAnno_df<-peakAnno_distal
# Check if peakAnno_df contains the expected data
print(head(peakAnno_df))

# Check the distribution of custom categories
print(table(peakAnno_df$customCategory))

# Calculate the percentage of each category
custom_anno_counts <- table(peakAnno_df$customCategory)
print(custom_anno_counts)

# Create custom_anno_df
custom_anno_df <- data.frame(
  Category = names(custom_anno_counts),
  Count = as.numeric(custom_anno_counts)
)
custom_anno_df$Percentage <- (custom_anno_df$Count / sum(custom_anno_df$Count)) * 100
print(custom_anno_df)

# Specify the order of categories for plotting
category_order <- c("Distal Intergenic", "Genes", "MSR", "SINEs", "LINEs","LTRs")

# Filter the data frame to include only the specified categories and in the specified order
custom_anno_df <- custom_anno_df[custom_anno_df$Category %in% category_order, ]

# Reorder the factor levels of Category column to match the specified order
custom_anno_df$Category <- factor(custom_anno_df$Category, levels = category_order)

g<-ggplot(custom_anno_df, aes(x = "", y = Percentage, fill = paste0(Category, " (", round(Percentage, 1), "%)"))) +
  geom_col(width = 1, position = "fill") +
  coord_polar("y", start = 0) +
  theme_void() +
  theme(
    legend.position = "right",
    plot.title = element_text(hjust = 0.5)  # Center the title
  ) +
  # scale_fill_brewer(palette = "Set3", direction = 1) + # You can change the color palette
  scale_fill_manual(values=c("black","grey", "darkred", "yellow",  "orange", "blue","green")) +
  labs(fill = "Category", title = "INTS11 peaks")

pdf(file="/data/jenuwein/processing/05.Alecia/20231212_ChipSeq_INTS11/20250227_INTS11_pieChart.pdf")
g
dev.off()
```

##Plot a pie chart for different reapeats categories

```{r, echo=FALSE}
##################
#extract only the distal intergenic peaks that are not annotated by anything else

peaks <- readPeakFile("/data/jenuwein/processing/05.Alecia/code/AleciasProject/INTS11_mergedPeaks.bed")

peakAnno <- annotatePeak(peaks, tssRegion=c(-3000, 3000),
                         TxDb=txdb, annoDb="TxDb.Mmusculus.UCSC.mm10.ensGene")

plotAnnoPie(peakAnno)


# Convert to data frames
peakAnno_df <- as.data.frame(peakAnno)

peakAnno_distal<-peakAnno_df[peakAnno_df$annotation == "Distal Intergenic",]

LINE_repeats <- import.bed("/data/jenuwein/processing/05.Alecia/code/LINEs.5UTR.filtered.bed")

peaks_gr <- GRanges(seqnames = peakAnno_distal$seqnames, ranges = IRanges(peakAnno_distal$start, peakAnno_distal$end))

# Overlap with repeats only for peaks not annotated as LINEs
repeats_gr <- GRanges(seqnames = seqnames(LINE_repeats), ranges = IRanges(start(LINE_repeats), end(LINE_repeats)))
repeat_hits <- findOverlaps(peaks_gr, repeats_gr)
peakAnno_distal$customCategory[queryHits(repeat_hits)] <- "LINEs"

#SINEs
SINE_repeats <- import.bed("/data/jenuwein/processing/05.Alecia/code/SINE.mm10.bed")
repeats_gr <- GRanges(seqnames = seqnames(SINE_repeats), ranges = IRanges(start(SINE_repeats), end(SINE_repeats)))
repeat_hits <- findOverlaps(peaks_gr, repeats_gr)
peakAnno_distal$customCategory[queryHits(repeat_hits)] <- "SINEs"

#LTR
LTR_repeats <- import.bed("/data/jenuwein/processing/05.Alecia/code/LTR.mm10.bed")
repeats_gr <- GRanges(seqnames = seqnames(LTR_repeats), ranges = IRanges(start(LTR_repeats), end(LTR_repeats)))
repeat_hits <- findOverlaps(peaks_gr, repeats_gr)
peakAnno_distal$customCategory[queryHits(repeat_hits)] <- "LTRs"


peakAnno_df<-peakAnno_distal
# Check if peakAnno_df contains the expected data
print(head(peakAnno_df))

# Check the distribution of custom categories
print(table(peakAnno_df$customCategory))

# Calculate the percentage of each category
custom_anno_counts <- table(peakAnno_df$customCategory)
print(custom_anno_counts)

#L1Md_A
L1Md_A_repeats <- import.bed("/data/jenuwein/processing/05.Alecia/code/L1Md_A.5UTR.bed")
repeats_gr <- GRanges(seqnames = seqnames(L1Md_A_repeats), ranges = IRanges(start(L1Md_A_repeats), end(L1Md_A_repeats)))
repeat_hits <- findOverlaps(peaks_gr, repeats_gr)
peakAnno_distal$customCategory[queryHits(repeat_hits)] <- "L1Md_A"

#L1Md_T
L1Md_T_repeats <- import.bed("/data/jenuwein/processing/05.Alecia/code/L1Md_T.5UTR.bed")
repeats_gr <- GRanges(seqnames = seqnames(L1Md_T_repeats), ranges = IRanges(start(L1Md_T_repeats), end(L1Md_T_repeats)))
repeat_hits <- findOverlaps(peaks_gr, repeats_gr)
peakAnno_distal$customCategory[queryHits(repeat_hits)] <- "L1Md_T"

ERVs_repeats <- import.bed("/data/jenuwein/processing/05.Alecia/code/ERVs.bed")
repeats_gr <- GRanges(seqnames = seqnames(ERVs_repeats), ranges = IRanges(start(ERVs_repeats), end(ERVs_repeats)))
repeat_hits <- findOverlaps(peaks_gr, repeats_gr)
peakAnno_distal$customCategory[queryHits(repeat_hits)] <- "LTRs/ERVs"

#GSAT
GSAT_repeats <- import.bed("/data/jenuwein/processing/05.Alecia/code/GSAT.mm10.bed")
repeats_gr <- GRanges(seqnames = seqnames(GSAT_repeats), ranges = IRanges(start(GSAT_repeats), end(GSAT_repeats)))
repeat_hits <- findOverlaps(peaks_gr, repeats_gr)
peakAnno_distal$customCategory[queryHits(repeat_hits)] <- "MSR"

peakAnno_df<-peakAnno_distal
# Check if peakAnno_df contains the expected data
print(head(peakAnno_df))

peakAnno_df2 <-peakAnno_df[peakAnno_df$customCategory != "Distal Intergenic",]

peakAnno_df_bk<-peakAnno_df
peakAnno_df<-peakAnno_df2 

print(head(peakAnno_df))

# Check the distribution of custom categories
print(table(peakAnno_df$customCategory))

# Calculate the percentage of each category
custom_anno_counts <- table(peakAnno_df$customCategory)
print(custom_anno_counts)

# Create custom_anno_df
custom_anno_df <- data.frame(
  Category = names(custom_anno_counts),
  Count = as.numeric(custom_anno_counts)
)
custom_anno_df$Percentage <- (custom_anno_df$Count / sum(custom_anno_df$Count)) * 100
print(custom_anno_df)

# Specify the order of categories for plotting
category_order <- c("Distal Intergenic", "MSR", "SINEs", "LINEs","L1Md_A", "L1Md_T", "LTRs", "LTRs/ERVs")


# Filter the data frame to include only the specified categories and in the specified order
custom_anno_df <- custom_anno_df[custom_anno_df$Category %in% category_order, ]

# Reorder the factor levels of Category column to match the specified order
custom_anno_df$Category <- factor(custom_anno_df$Category, levels = category_order)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
