{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f6d2cfad",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_line_breaks(input_file, output_file):\n",
    "    # Read the contents of the input file\n",
    "    with open(input_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "\n",
    "    # Remove incorrect line breaks\n",
    "    corrected_lines = []\n",
    "    current_sequence = \"\"\n",
    "    for line in lines:\n",
    "        if line.startswith(\">\"):\n",
    "            # If a new sequence header is encountered, append the previous sequence (with corrected line breaks)\n",
    "            if current_sequence:\n",
    "                corrected_lines.append(current_sequence)\n",
    "            corrected_lines.append(line)\n",
    "            current_sequence = \"\"\n",
    "        else:\n",
    "            # Remove line breaks within the sequence data\n",
    "            current_sequence += line.strip()\n",
    "\n",
    "    # Append the last sequence (with corrected line breaks)\n",
    "    if current_sequence:\n",
    "        corrected_lines.append(current_sequence)\n",
    "\n",
    "    # Write the corrected lines to the output file\n",
    "    with open(output_file, 'w') as file:\n",
    "        file.writelines(corrected_lines)\n",
    "\n",
    "#fix_line_breaks('/data/jenuwein/processing/20230412_Alecia_PoreC/chr2_withInsertion.fa', '/data/jenuwein/processing/20230412_Alecia_PoreC/chr2_withInsertion.fixed.fa')\n",
    "\n",
    "fix_line_breaks('/data/jenuwein/processing/20230412_Alecia_PoreC/mm10.PlusInsertion.fa', '/data/jenuwein/processing/20230412_Alecia_PoreC/mm10.PlusInsertion.fixed.fa')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "41838ca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def exchange_sequences(input_file, sequence_name1, sequence_file2, sequence_name2, output_file):\n",
    "    # Read the contents of the input file\n",
    "    with open(input_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "\n",
    "    # Find the index of the line corresponding to sequence 1\n",
    "    index1 = None\n",
    "    for i, line in enumerate(lines):\n",
    "        if line.startswith(\">\" + sequence_name1):\n",
    "            index1 = i\n",
    "            break\n",
    "\n",
    "    if index1 is None:\n",
    "        print(\"Sequence 1 not found in the input file.\")\n",
    "        return\n",
    "\n",
    "    # Read the contents of sequence file 2\n",
    "    with open(sequence_file2, 'r') as file:\n",
    "        sequence2_lines = file.readlines()\n",
    "\n",
    "    # Create the sequence 2 entry with the new name and sequence\n",
    "    sequence2_entry = [\">\" + sequence_name2 + \"\\n\"] + sequence2_lines[1:]\n",
    "\n",
    "    # Replace the sequence 1 entry with sequence 2 entry\n",
    "    lines[index1:index1 + 2] = sequence2_entry\n",
    "\n",
    "    # Write the modified lines to the output file\n",
    "    with open(output_file, 'w') as file:\n",
    "        file.writelines(lines)\n",
    "\n",
    "# Usage example\n",
    "exchange_sequences('/data/repository/organisms/GRCm38_ensembl/genome_fasta/genome.fa', '2 dna_sm:chromosome chromosome:GRCm38:2:1:182113224:1 REF', '/data/jenuwein/processing/20230412_Alecia_PoreC/chr2.withInsertFinal.fa', '2', '/data/jenuwein/processing/20230412_Alecia_PoreC/mm10.PlusInsertion.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "de50f191",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "MT\n",
      "X\n",
      "Y\n",
      "JH584299.1\n",
      "GL456233.1\n",
      "JH584301.1\n",
      "GL456211.1\n",
      "GL456350.1\n",
      "JH584293.1\n",
      "GL456221.1\n",
      "JH584297.1\n",
      "JH584296.1\n",
      "GL456354.1\n",
      "JH584294.1\n",
      "JH584298.1\n",
      "JH584300.1\n",
      "GL456219.1\n",
      "GL456210.1\n",
      "JH584303.1\n",
      "JH584302.1\n",
      "GL456212.1\n",
      "JH584304.1\n",
      "GL456379.1\n",
      "GL456216.1\n",
      "GL456393.1\n",
      "GL456366.1\n",
      "GL456367.1\n",
      "GL456239.1\n",
      "GL456213.1\n",
      "GL456383.1\n",
      "GL456385.1\n",
      "GL456360.1\n",
      "GL456378.1\n",
      "GL456389.1\n",
      "GL456372.1\n",
      "GL456370.1\n",
      "GL456381.1\n",
      "GL456387.1\n",
      "GL456390.1\n",
      "GL456394.1\n",
      "GL456392.1\n",
      "GL456382.1\n",
      "GL456359.1\n",
      "GL456396.1\n",
      "GL456368.1\n",
      "JH584292.1\n"
     ]
    }
   ],
   "source": [
    "#split fasta\n",
    "import os\n",
    "\n",
    "def split_fasta(input_file, output_directory):\n",
    "    # Read the contents of the input file\n",
    "    with open(input_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "\n",
    "    # Initialize variables\n",
    "    sequence_name = None\n",
    "    sequence_lines = []\n",
    "\n",
    "    # Iterate over each line in the input file\n",
    "    for line in lines:\n",
    "        if line.startswith(\">\"):\n",
    "            # If a new sequence header is encountered, write the previous sequence to a file\n",
    "            if sequence_name is not None:\n",
    "                sequence_name = sequence_name.split(' ')[0]\n",
    "                print(sequence_name)\n",
    "                write_sequence_file(sequence_name, sequence_lines, output_directory)\n",
    "\n",
    "            # Extract the sequence name from the header\n",
    "            sequence_name = line[1:].split(' ')[0]\n",
    "            sequence_lines = []\n",
    "        else:\n",
    "            # Collect the lines of the current sequence\n",
    "            sequence_lines.append(line)\n",
    "\n",
    "    # Write the last sequence to a file\n",
    "    if sequence_name is not None:\n",
    "        write_sequence_file(sequence_name, sequence_lines, output_directory)\n",
    "\n",
    "def write_sequence_file(sequence_name, sequence_lines, output_directory):\n",
    "    # Generate the filename based on the sequence name\n",
    "    filename = sequence_name + \".fasta\"\n",
    "\n",
    "    # Create the output directory if it doesn't exist\n",
    "    os.makedirs(output_directory, exist_ok=True)\n",
    "\n",
    "    # Write the sequence lines to the output file in the specified directory\n",
    "    output_file = os.path.join(output_directory, filename)\n",
    "    with open(output_file, 'w') as file:\n",
    "        file.write(\">\"+sequence_name+\"\\n\")\n",
    "        file.write(\"\".join(sequence_lines))\n",
    "\n",
    "# Usage example\n",
    "split_fasta('/data/repository/organisms/GRCm38_ensembl/genome_fasta/genome.fa', '/data/jenuwein/processing/20230412_Alecia_PoreC/split_fasta/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5a092bf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#change line length\n",
    "def change_line_length(input_file, output_file, line_length=60):\n",
    "    # Read the contents of the input file\n",
    "    with open(input_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "\n",
    "    # Remove existing line breaks and create new lines with the desired line length\n",
    "    sequence=\"\"\n",
    "    new_lines = []\n",
    "    for line in lines:\n",
    "        if line.startswith(\">\"):\n",
    "            # Header line, no modification needed\n",
    "            new_lines.append(line)\n",
    "        else:\n",
    "            # Sequence line, remove line breaks and create new lines with the desired length\n",
    "            sequence = sequence+line.strip()\n",
    "                \n",
    "    for i in range(0, len(sequence), line_length):\n",
    "        new_lines.append(sequence[i:i+line_length] + '\\n')\n",
    "\n",
    "    # Write the modified lines to the output file\n",
    "    with open(output_file, 'w') as file:\n",
    "        file.writelines(new_lines)\n",
    "\n",
    "# Usage example\n",
    "change_line_length('/data/jenuwein/processing/20230412_Alecia_PoreC/customeGenome/chr2_plusInsertion.fa', '/data/jenuwein/processing/20230412_Alecia_PoreC/customeGenome/chr2_plusInsertion.FINAL.fa', line_length=60)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "bc03c400",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "continue\n",
      "we got problems\n",
      "3035291\n"
     ]
    }
   ],
   "source": [
    "def check_length(input_file):\n",
    "    # Read the contents of the input file\n",
    "    with open(input_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "\n",
    "    # Remove existing line breaks and create new lines with the desired line length\n",
    "    sequence=\"\"\n",
    "    new_lines = []\n",
    "    counter=0\n",
    "    for line in lines:\n",
    "        counter=counter+1\n",
    "        if line.startswith(\">\"):\n",
    "            print(\"continue\")\n",
    "        else:\n",
    "            if len(line.strip()) != 60:\n",
    "                print(\"we got problems\")\n",
    "                print(str(counter))\n",
    "                break\n",
    "\n",
    "#check_length('/data/jenuwein/processing/20230412_Alecia_PoreC/customeGenome/chr2_plusInsertion.FINAL.fa')\n",
    "check_length('/data/jenuwein/processing/20230412_Alecia_PoreC/split_fasta/2.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f250e956",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "PoreC",
   "language": "python",
   "name": "porec"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
